From ddf3e3509944e07be07ed17ce6dd3547a34cdb9b Mon Sep 17 00:00:00 2001
From: samaga Prasanna Krishna <samaga.krishna@intel.com>
Date: Wed, 7 Nov 2012 12:57:40 +0530
Subject: [PATCH 070/170] [PORT FROM MAIN][CTP-Offload] TPV DoubleTwist
 Player: Cannot play music.

BZ: 62959

This patch will fix the offload playback issue in DoubleTwist player.

During offload track creation, active track count in effect chain is
not getting incremented due to track's main buffer and mMixBuffer is same.
Hence extra check is needed to check for effect chain and increment the
active track count.

Change-Id: I6fb3c643c8a568b8aa2b2614b4cd8ad5f7512b5f
origChange-Id: Ib84e0fd1bf5054ea59ea530937f71cc762082e94
Signed-off-by: samaga Prasanna Krishna <samaga.krishna@intel.com>
Reviewed-on: http://android.intel.com:8080/77891
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Tested-by: M, Arulselvan <arulselvan.m@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 services/audioflinger/AudioFlinger.cpp |   18 ++++++++++++++++++
 1 files changed, 18 insertions(+), 0 deletions(-)

diff --git a/services/audioflinger/AudioFlinger.cpp b/services/audioflinger/AudioFlinger.cpp
index 68f3818..fe09aa7 100644
--- a/services/audioflinger/AudioFlinger.cpp
+++ b/services/audioflinger/AudioFlinger.cpp
@@ -1997,6 +1997,24 @@ status_t AudioFlinger::PlaybackThread::addTrack_l(const sp<Track>& track)
         track->mResetDone = false;
         track->mPresentationCompleteFrames = 0;
         mActiveTracks.add(track);
+
+#ifdef INTEL_MUSIC_OFFLOAD_FEATURE
+        // The check track->mainBuffer() != mMixBuffer to detect effect chain
+        // is connected to the track may not always work in case of offload
+        // track because track main buffer will be same as mMixBuffer.
+        // This will happen if effects are created before this new track is
+        // created and hence we need to get the effect chain.
+        if (isOffloadTrack()) {
+            sp<EffectChain> chain = getEffectChain_l(track->sessionId());
+            if ((track->mainBuffer() != mMixBuffer) || (chain != 0)) {
+                ALOGV("addTrack_l() starting track on chain %p for session %d", chain.get(), track->sessionId());
+                chain->incActiveTrackCnt();
+            }
+
+            mWaitWorkCV.broadcast();
+            return NO_ERROR;
+        }
+#endif
         if (track->mainBuffer() != mMixBuffer) {
             sp<EffectChain> chain = getEffectChain_l(track->sessionId());
             if (chain != 0) {
-- 
1.7.9

